content/ui-patterns/forms.mdx

```diff
@@ -0,0 +1,393 @@
+
+### Inline validation
+
+<Note variant="warning">
+  Inline validation offers immediate feedback, but be sure to consider the following drawbacks.
+
+  <ul>
+    <li>There is a negative impact on perceived performance when validating server-side.</li>
+    <li>When a screen reader user moves focus from the invalid input to the next form control, they will be interrupted by the validation message of the previous form control.</li>
+  </ul>
+</Note>
+
+<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
+    <img
+        width="456"
+        alt="screen recording of validation being performed after the user changes and blurs"
+        src="https://user-images.githubusercontent.com/2313998/164065853-fd7ef55a-7082-4b0d-aa6b-2b25af4c664b.gif"
+    />
+    <Box as="p" mt="0">If the form control is in a valid state, validation should be performed until after the user has made a change to the input <strong>and has removed focus from the input</strong>.</Box>
+</Box>
+
+<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
+    <img
+        width="456"
+        alt="screen recording of validation being performed as the user tries to fix an invalid input"
+        src="https://user-images.githubusercontent.com/2313998/164065860-25e23fa0-e22a-435d-adbd-5666a9fd7890.gif"
+    />
+    <Box as="p" mt="0">Don't attempt to validate an input before the user is done with it. Validation may be performed as the user is typing or making their selection, but only <strong>after the first time the input has been validated</strong> and the <strong>input is in an invalid state</strong>. This gives the user early positive feedback by removing the error if the user makes a change that fixes it.</Box>
+</Box>
+
+<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
+    <img
+        width="456"
+        alt="screen recording of valid repo name input showing loading indicator after blur"
+        src="https://user-images.githubusercontent.com/2313998/164065857-573ca6db-f2c7-49ac-a3cf-bf479a40973d.gif"
+    />
+    <Box as="p" mt="0">If the form control’s validation is likely to take more than 1 second, show a loading indicator.</Box>
+</Box>
+
+### Validating a group of form controls
+
+<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
+    <img
+        width="456"
+        alt="checkbox group with an error"
+        src="https://user-images.githubusercontent.com/2313998/164065224-d1ba56ce-c632-4318-b9b2-b75940f771a6.png"
+    />
+    <Box as="p" mt="0">A validation message may be used to describe an error for a group of form controls. For example: if a user checks 3 checkboxes when they’re only allowed to check 2 options.</Box>
```

Comment:
I think my comment above about grouping error messages could be described in this section instead. The case you mention here is a single error message for a group of controls, but we should document how to present multiple individual error messages for a group of controls as well.

Reply:
Here's a screenshot from a previous form I worked on:

![Contact form with multiple fields showing an error state. At the top of the form, there's a group with a title and list of links to each field, stating what the error is with the field](https://user-images.githubusercontent.com/18661030/168659393-55b9a577-3dee-44e8-b2d1-bf03ff326cfb.png)

I think the idea here is if there are multiple errors in a single `<form>`, show them in a labeled group with links to each field. My original comment might have sounded like multiple errors per _field_, but that's not what I mean 😅 that feels problematic in a different way!

@alliethu any thoughts here? I pulled this from a book on accessible forms, and that was a year ago. I haven't researched this recently.

Reply:
@langermank Your annotation is very much inline with what I would have done myself. I'm attaching a rough annotation below that includes a few more details: 

![Error summary notation](https://user-images.githubusercontent.com/7228610/168684134-1c58188a-e8af-4349-bcbc-cf04cbeedb1a.png)

Considerations to make:
- add `role="alert"` to the summary `div`
- move keyboard focus to the error summary
- use a semantic heading (I'm thinking an `h2` to title the error summary area)
- reference the heading in the summary `div` by using `aria-labelledby`
- use a semantic list for the error summaries
- when a link is activated, place focus in its corresponding input field
- the error messages should be worded the same in the error summary and inline with each input.
cc/ @mperrotti

Reply:
Thank you both for the helpful examples! I misunderstood what was being suggested. I thought @langermank was suggesting this pattern for a single group of checkbox or radio inputs.

It feels aggressive to _always_ show the error summary alert banner. We should explain when a summary is needed and when it isn't. Maybe when there are 3 or more errors?

Reply:
I'm noticing repetitive error messages in @alliethu's example, and that feels like it could be overwhelming or confusing to sighted and non-sighted users.

Reply:
I pushed [a commit](https://github.com/primer/design/pull/261/commits/0c55099ae9f157dbdc2c9857d1e1070638ecbdf6) to add some guidance on designing error summaries. I'm not feeling great about it yet, and I'd appreciate some time pairing with somebody from the accessibility team.